File: apps/mobile/modules/novella-readium/ios/NovellaReadiumModule.swift
import ExpoModulesCore

public final class NovellaReadiumModule: Module {
  public func definition() -> ModuleDefinition {
    Name("NovellaReadium")

    View(NovellaReadiumView.self) {
      ViewName("Reader")

      Prop("publicationUri") { (view: NovellaReadiumView, value: String) in
        view.setPublicationUri(value)
      }
      Prop("publicationId") { (view: NovellaReadiumView, value: String) in
        view.setPublicationId(value)
      }
      Prop("declaredHrefs") { (view: NovellaReadiumView, value: [String]) in
        view.setDeclaredHrefs(value)
      }
      Prop("initialLocator") { (view: NovellaReadiumView, value: [String: Any]?) in
        view.setInitialLocator(value)
      }
      Prop("preferences") { (view: NovellaReadiumView, value: [String: Any]) in
        view.setPreferences(value)
      }
      Prop("contentInsets") { (view: NovellaReadiumView, value: [String: Double]) in
        view.setContentInsets(value)
      }

      Events("onReady", "onLocatorChange", "onLink", "onImage", "onError", "onStatus")

      AsyncFunction("getCurrentLocator") { (view: NovellaReadiumView) async -> [String: Any]? in
        await view.getCurrentLocator()
      }
      AsyncFunction("goToLocator") { (view: NovellaReadiumView, locator: [String: Any]) async throws -> Bool in
        try await view.goToLocator(locator)
      }
      AsyncFunction("goForward") { (view: NovellaReadiumView) async -> Bool in
        await view.goForward()
      }
      AsyncFunction("goBackward") { (view: NovellaReadiumView) async -> Bool in
        await view.goBackward()
      }
    }
  }
}

File: apps/mobile/modules/novella-readium/android/src/main/java/sh/celia/novella/modules/readium/NovellaReadiumModule.kt
package sh.celia.novella.modules.readium

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class NovellaReadiumModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("NovellaReadium")

    View(ComposeReadiumView::class) {
      Name("Reader")

      Prop("publicationUri") { view: ComposeReadiumView, value: String ->
        view.setPublicationUri(value)
      }
      Prop("publicationId") { view: ComposeReadiumView, value: String ->
        view.setPublicationId(value)
      }
      Prop("declaredHrefs") { view: ComposeReadiumView, value: List<String> ->
        view.setDeclaredHrefs(value)
      }
      Prop("initialLocator") { view: ComposeReadiumView, value: Map<String, Any>? ->
        view.setInitialLocator(value)
      }
      Prop("preferences") { view: ComposeReadiumView, value: Map<String, Any> ->
        view.setPreferences(value)
      }
      Prop("contentInsets") { view: ComposeReadiumView, value: Map<String, Double> ->
        view.setContentInsets(value)
      }

      Events("onReady", "onLocatorChange", "onLink", "onError")

      AsyncFunction("goToLocator") { view: ComposeReadiumView, locator: Map<String, Any> ->
        view.goToLocator(locator)
      }
      AsyncFunction("goForward") { view: ComposeReadiumView ->
        view.goForward()
      }
      AsyncFunction("goBackward") { view: ComposeReadiumView ->
        view.goBackward()
      }
    }
  }
}
